File: NotificationCenter/UpNotificationCenter/UpNotificationCenter/UpNotificationCenter.m
//
//  UpNotificationCenter.m
//  UpNotificationCenter
//
//  Created by Chauyan Wang on 10/14/16.
//  Copyright © 2016 upshotech. All rights reserved.
//

#import "UpNotificationCenter.h"
#import "UpObserver.h"

@implementation UpNotificationCenter
{
    NSMutableArray *observerArray;
}

- (instancetype) init {
    if (self = [super init]) {
        observerArray = [[NSMutableArray alloc] init];
    }
    return self;
}

+ (UpNotificationCenter *)defaultCenter {
    static UpNotificationCenter *defaultCenter;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        defaultCenter = [[self alloc] init];
    });

    return defaultCenter;
}

- (void) addObserver:(id)observer selector:(SEL)aSelector name:(NSString *)notificationName object:(id)userObject {
    UpObserver *model = [[UpObserver alloc] init];
    model.observer = observer;
    model.userObject = userObject;
    model.selector = aSelector;
    model.notificationName = notificationName;
    [observerArray addObject:model];
}

- (void) postNotificationName:(NSString *)notificationName object:(id)userObject {
    [observerArray enumerateObjectsUsingBlock:^(UpObserver* obj, NSUInteger idx, BOOL * _Nonnull stop) {
        if ([obj.notificationName isEqualToString:notificationName]) {
            [obj.observer performSelector:obj.selector withObject:obj.userObject];
        }
    }];
}

- (void) removeObserver:(id)observer {
    __block UpObserver *model;
    [observerArray enumerateObjectsUsingBlock:^(UpObserver* obj, NSUInteger idx, BOOL * _Nonnull stop) {
        if ([obj.observer isKindOfClass:[observer class]]) {
            model = obj;
        }
    }];
    
    [observerArray removeObject:model];
}
@end

File: NotificationCenter/UpNotificationCenter-Swift/UpNotificationCenter-Swift/UpNotificationCenter.swift
//
//  UpNotificationCenter.swift
//  UpNotificationCenter-Swift
//
//  Created by Chauyan Wang on 10/14/16.
//  Copyright © 2016 upshotech. All rights reserved.
//

import UIKit

open class UpNotificationCenter: NSObject {
    
    let observerArray = NSMutableArray()
    static let selfInstnace = UpNotificationCenter()
    
    open class func defaultCenter() -> UpNotificationCenter {
        return selfInstnace
    }
    
    open func addObserver(_ observer: Any, selector aSelector: Selector, name aName: NSString?, object anObject: Any?) {
        let model = UpObserver()
        model.initObserver(observer, selector: aSelector, notificationName: aName, object: anObject)
        
        observerArray.add(model)
    }
    
    open func post(name aName: NSString, object anObject: Any?) {
        observerArray.enumerateObjects ({ (obj, index, stop) in
            let model = obj as! UpObserver

            if (model.notificationName!.isEqual(to: aName as String)) {
                
                (model.obsever as! NSObject).perform(model.selector, with: model.userObject)
            }
        })
    }
    
    open func removeObserver(_ observer: Any) {
        var model:UpObserver? = nil
        observerArray.enumerateObjects ({ (obj, index, stop) in
            model = obj as? UpObserver
            
            if (model?.obsever as AnyObject === obj as AnyObject) {
                return
            }
        })
        
        if (model != nil) {
            observerArray.remove(model)
        }
    }

}
